lib/libopenswan/optionsfrom.3
.\"     Title: IPSEC_OPTIONSFROM
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.73.2 <http://docbook.sf.net/>
.\"      Date: 11/14/2008
.\"    Manual: 16 Oct 1998
.\"    Source: 16 Oct 1998
.\"
.TH "IPSEC_OPTIONSFROM" "3" "11/14/2008" "16 Oct 1998" "16 Oct 1998"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
ipsec optionsfrom - read additional ``command-line'' options from file
.SH "SYNOPSIS"
.sp
.ft B
.nf
#include <freeswan\.h>

.fi
.ft
.HP 24
.BI "const char *optionsfrom(char\ *\ " "filename" ", int\ *\ " "argcp" ", char\ ***\ " "argvp" ", int\ " "optind" ", FILE\ *\ " "errsto" ");"
.SH "DESCRIPTION"
.PP
\fIOptionsfrom\fR
is called from within a
\fBgetopt_long\fR(3)
scan, as the result of the appearance of an option (preferably
\fB\-\-optionsfrom\fR) to insert additional \(lqcommand\-line\(rq arguments into the scan immediately after the option\. Typically this would be done to pick up options which are security\-sensitive and should not be visible to
\fBps\fR(1)
and similar commands, and hence cannot be supplied as part of the actual command line or the environment\.
.PP
\fIOptionsfrom\fR
reads the additional arguments from the specified
\fIfilename\fR, allocates a new argument vector to hold pointers to the existing arguments plus the new ones, and amends
\fIargc\fR
and
\fIargv\fR
(via the pointers
\fIargcp\fR
and
\fIargvp\fR, which must point to the
\fIargc\fR
and
\fIargv\fR
being supplied to
\fBgetopt_long\fR(3)) accordingly\.
\fIOptind\fR
must be the index, in the original argument vector, of the next argument\.
.PP
If
\fIerrsto\fR
is NULL,
\fBoptionsfrom\fR
returns NULL for success and a pointer to a string\-literal error message for failure; see DIAGNOSTICS\. If
\fIerrsto\fR
is non\-NULL and an error occurs,
\fBoptionsfrom\fR
prints a suitable complaint onto the
\fIerrsto\fR
descriptor and invokes
\fIexit\fR
with an exit status of 2; this is a convenience for cases where more sophisticated responses are not required\.
.PP
The text of existing arguments is not disturbed by
\fBoptionsfrom\fR, so pointers to them and into them remain valid\.
.PP
The file of additional arguments is an ASCII text file\. Lines consisting solely of white space, and lines beginning with
\fB#\fR, are comments and are ignored\. Otherwise, a line which does not begin with
\fB\-\fR
is taken to be a single argument; if it both begins and ends with double\-quote ("), those quotes are stripped off (note, no other processing is done within the line!)\. A line beginning with
\fB\-\fR
is considered to contain multiple arguments separated by white space\.
.PP
Because
\fBoptionsfrom\fR
reads its entire file before the
\fBgetopt_long\fR(3)
scan is resumed, an
\fBoptionsfrom\fR
file can contain another
\fB\-\-optionsfrom\fR
option\. Obviously, infinite loops are possible here\. If
\fIerrsto\fR
is non\-NULL,
\fBoptionsfrom\fR
considers it an error to be called more than 100 times\. If
\fIerrsto\fR
is NULL, loop detection is up to the caller (and the internal loop counter is zeroed out)\.
.SH "EXAMPLE"
.PP
A reasonable way to invoke
\fBoptionsfrom\fR
would be like so:
.sp
.RS 4
.nf
#include <getopt\.h>

struct option opts[] = {
	/* \.\.\. */
	"optionsfrom",	1,	NULL,	\'+\',
	/* \.\.\. */
};

int
main(argc, argv)
int argc;
char *argv[];
{
	int opt;
	extern char *optarg;
	extern int optind;

	while ((opt = getopt_long(argc, argv, "", opts, NULL)) != EOF)
		switch (opt) {
		/* \.\.\. */
		case \'+\':	/* optionsfrom */
			optionsfrom(optarg, &argc, &argv, optind, stderr);
			/* does not return on error */
			break;
		/* \.\.\. */
		}
	/* \.\.\. */
.fi
.RE
.sp
.SH "SEE ALSO"
.PP
\fBgetopt_long\fR(3)
.SH "DIAGNOSTICS"
.PP
Errors in
\fBoptionsfrom\fR
are: unable to open file; attempt to allocate temporary storage for argument or argument vector failed; read error in file; line too long\.
.SH "HISTORY"
.PP
Written for the FreeS/WAN project by Henry Spencer\.
.SH "BUGS"
.PP
The double\-quote convention is rather simplistic\.
.PP
Line length is currently limited to 1023 bytes, and there is no continuation convention\.
.PP
The restriction of error reports to literal strings (so that callers don\'t need to worry about freeing them or copying them) does limit the precision of error reporting\.
.PP
The error\-reporting convention lends itself to slightly obscure code, because many readers will not think of NULL as signifying success\.
.PP
There is a certain element of unwarranted chumminess with the insides of
\fBgetopt_long\fR(3)
here\. No non\-public interfaces are actually used, but
\fBoptionsfrom\fR
does rely on
\fBgetopt_long\fR(3)
being well\-behaved in certain ways that are not actually promised by the specs\.


CHANGES

Note that listings are in chronological order of release times, not in order
of version numbers, so you will find 2.5.x and 2.6.x releases intersperced.

See also docs/KNOWN_BUGS and http://bugs.openswan.org/

v2.6.19
* Fix for ERROR: Module xfrm6_tunnel is in use by ipcomp6 [paul]
* Fix for using MODPROBE=insmod where insmod does not support -q [paul]
* Enable all wins/dns options as specified in man pages [david]
* build support for all WINS/DNS options as mentioned in
  the man pages [david]
* Removed obsolete keywords: firewall (linux 2.0), spibase, spi,espenckey,
  espauthkey and espreplay_window (manual keying) [paul]
* Fix unneccesary and bogus connection switching with NAT [Shingo Yamawaki]
  (this might relate to several reported bugs in the tracker)
* Added cisco-decrypt utility for PCF obfuscation in contrib/ [paul]
* Fix for crasher when the leftcert= filename was not found [paul]
* Patch for "route already in use" when using two different IP's
  to talk to the same remote IP using two tunnels [Avesh Agarwal]
* Fixes to init scripts [Avesh/Tuomo]
  See also: RedHat bugzilla #466861.
* Bugtracker bugs fixed:
  #992: keyingretries default changed from %forever to 3 [ken]
        (bug was introduced in 2.6.x)
  #981: plutodebug=all klipsdebug=all not operate. [paul]

v2.6.18
* Fix for compiling KLIPS on RHEL/Centos 2.6.18-92.1.10.el5 [dhr/paul]
* Fix in deleting connections that might have caused some of our Delete
  Notifies to have gotten lost. Introduced in openswan 2.5.01 [paul]
* Rekey= inverted yes/no, causing rekey=no to be rekey=yes [Shingo Yamawaki]
* Some memory leaks / refcount fixes [Shingo Yamawaki]
* Removed most of #ifdef CONFIG_KLIPS_DEBUG conditionals. We now always
  compile in DEBUG support. [paul]
* No longer use the assembly version of des_encrypt (dx86unix.S). It
  is i386-i686 specific, requires framepointers and does not work with
  CONFIG_REGPARM=y, which is the unconditional default for 2.6.17+ [paul]
* Fix memory leak when we run out of descriptors [David McCullough]
* Various memory leak fixes for pluto (from #macosx) [Ilia Sotnikov]
* LEAK_DETECTIVE should report better now [Ilia Sotnikov]
* Add support for USE_DMALLOC [Ilia Sotnikov]
* Update stats to show dropped packets [David McCullough]
* Allow session migration of OCF devices [Brad Vrabete]
* DNS/WINS ModeConfig fixes [David McCullough]
* refineconnection bug fix. This might cover various problems where
  the right conn was not picked (eg rightca="%any" workaround, but
  perhaps also some rekey issues) [paul]
* unregister_netdevice: waiting for ppp2 to become free. Usage
  count = -1 on kernels < 2.6.24 [Martin Schiller]
* Fix for parallel building, eg with rpmbuild [tuomo]
* Bugtracker bugs fixed:
  #989: Patch for fixing type-punned compiler warnings [Alin Nastac]
  #979: Two errors in debian/ packaging files (fix included) [ruben]
  #978: ipsec.conf man page has typo in virtual_private sample line [tuomo]
  #975: ipsec_setup: Unknown socket write error 96. [paul]
  #231: In Aggressive Mode with NAT-T,initiator should switch port [hiren joshi]
  #228: Problems with %any matching in ipsec.secrets? [David McCullough]
  #984: OpenSwan 2.4.13: Wrong ipsec_dev_get(x) function for Kernels < 2.6.24 

v2.6.17 [will be skipped due to bad tag]

v2.6.16
* Merged in David McCullough's OCF patch [david/paul]
  Requires kernel patch, see http://ocf-linux.sourceforge.net/
* dpdaction=restart_by_peer support added [david]
* dynamic dns support (do dns lookup at restart conn, eg after dpd) [david]
  Uses USE_DYNAMICDNS=
* Added USE_SINGLE_CONF_DIR= [david]
* KLIPS support for 2.6.24 / 2.6.25 [david]
* Fix for "Unknown sysctl binary path" [david]
* rekeyfuzz is percentage, not integer [david]
* Added HAVE_STATSD= support to log state changes (for webgui etc) [david]
  (disabled per default)
* Wrapper to handle more then 2048 NETKEY tunnels [david]
* fixes for parser warnings [dhr]
* Fix rmmod calls not to use -s since busybox rmmod does not support it [paul]
* Fixes to KLIPS for newer 2.4 kernels [greg/davidm]
* Road Warrior behind NAT - Aggressive Mode: wrong NAT-T decision [hiren joshi]
* Added documentation for leftxauthusername= and XAUTH passwd support [paul]
* Bugtracker bugs fixed:
  #977: KLIPS doesn't work when wan interface is a tagged vlan interface
        Regression from 2.4? [Tino Keitel / Krisztian KOVACS]
  #972: Aggressive mode connection breaks after DPD timeout for NATed peer
  #965: xmlto man fails to generate ipsec.conf.5 man page  [tuomo]

v2.6.15
* Patch to support NETKEY backport on Debian kernels [Rene Mayrhofer]
* Fix a crasher when using right=%any with plutodebug=controlmore [paul]
* Fix a crasher when deleting connections in NETKEY [ken]
* Added disable_port_floating support to scripts and parser and
  repair the default back to allow port floating [paul]
* Change (back) defaults of plutorestartoncrash and uniqueids from
  no to yes. The new parser mistakenly did not set these [paul]
* Revert af family code in find_host_pair causing some connections to not
  be found in find_host_connection2() [paul]
* Fixes to _updown.mast, _realsetup (mast) and startklips [paul]
* Fixed to saref code so we can build on OSX again [paul]
* Use PREROUTING instead of OUTPUT/FORWARD for mast [mcr]
* NETKEY support for eroute_idle using get_sa_info() [herbert/andreas]
* Do not send DPD "R_Y_THERE" when eroute not idle [andreas]
* Support for Relative Distinguished Name "unstructuredName"/"UN"
  in ID_DER_ASN1_DN identities (eg leftid="UN=John Doe") [andreas]
* Removed forwardcontrol= and rp_filter= options. Ignore if present
  in config file. Use /etc/sysctl.conf [paul]
* Fix for left="%defaultroute" when using NETKEY [tuomo]
* Fix for KLIPS on SMP systems (missing SOCKOPS_WRAP for pfkey_ops) [dhr/paul]
* Merged in some IPsec SAref related code [mcr/paul]
* Merged in packaging/suse for building rpms on SLES [paul]
* Bugtracker bugs fixed:

lib/libopenswan/optionsfrom.c
/*
 * pick up more options from a file, in the middle of an option scan
 * Copyright (C) 1998, 1999  Henry Spencer.
 * 
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Library General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.  See <http://www.fsf.org/copyleft/lgpl.txt>.
 * 
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
 * License for more details.
 *
 * RCSID $Id: optionsfrom.c,v 1.9 2005/08/25 01:19:23 paul Exp $
 */
#include "internal.h"
#include "openswan.h"
#include "oswlog.h"

#include <stdio.h>

#define	MAXLOOP	100		/* loop-detection limit */

/* internal work area */
struct work {
#	define	LOTS	1024
	char buf[LOTS];
	char *line;
	char *pending;
};

static const char *dowork(const char *, int *, char ***, int);
static const char *getanarg(FILE *, struct work *, char **);
static char *of_getline(FILE *, char *, size_t);

/*
 - optionsfrom - add some options, taken from a file, to argc/argv
 * If errsto is non-NULL, does not return in event of error.
 */
const char *			/* NULL for success, else string literal */
optionsfrom(filename, argcp, argvp, optind, errsto)
const char *filename;
int *argcp;			/* pointer to argc */
char ***argvp;			/* pointer to argv */
int optind;			/* current optind, number of next argument */
FILE *errsto;			/* where to report errors (NULL means return) */
{
	const char *e;
	static int nuses = 0;

	if (errsto != NULL) {
		nuses++;
		if (nuses >= MAXLOOP) {
			fprintf(errsto,
				"%s: optionsfrom called %d times, looping?\n",
				(*argvp)[0], nuses);
			exit(2);
		}
	} else
		nuses = 0;

	e = dowork(filename, argcp, argvp, optind);
	if (e != NULL && errsto != NULL) {
		fprintf(errsto, "%s: optionsfrom failed: %s\n", (*argvp)[0], e);
		exit(2);
	}
	return e;
}

/*
 - dowork - do all the real work of optionsfrom
 * Does not alter the existing arguments, but does relocate and alter
 * the argv pointer vector.
 */
static const char *		/* NULL for success, else string literal */
dowork(filename, argcp, argvp, optind)
const char *filename;
int *argcp;			/* pointer to argc */
char ***argvp;			/* pointer to argv */
int optind;			/* current optind, number of next argument */
{
	char **newargv;
	char **tmp;
	int newargc;
	int next;		/* place for next argument */
	int room;		/* how many more new arguments we can hold */
#	define	SOME	10	/* first guess at how many we'll need */
	FILE *f;
	int i;
	const char *p;
	struct work wa;		/* for getanarg() */

	f = fopen(filename, "r");
	if (f == NULL)
		return "unable to open file";

	newargc = *argcp + SOME;
	newargv = malloc((newargc+1) * sizeof(char *));
	if (newargv == NULL)
		return "unable to allocate memory";
	memcpy(newargv, *argvp, optind * sizeof(char *));
	room = SOME;
	next = optind;

	newargv[next] = NULL;
	wa.pending = NULL;
	while ((p = getanarg(f, &wa, &newargv[next])) == NULL) {
		if (room == 0) {
			newargc += SOME;
			tmp = realloc(newargv, (newargc+1) * sizeof(char *));
			if (tmp == NULL) {
				p = "out of space for new argv";
				break;		/* NOTE BREAK OUT */
			}
			newargv = tmp;
			room += SOME;
		}
		next++;
		room--;
	}
	if (p != NULL && !feof(f)) {	/* error of some kind */
		for (i = optind+1; i <= next; i++)
			if (newargv[i] != NULL)
				free(newargv[i]);
		free(newargv);
		fclose(f);
		return p;
	}

	fclose(f);
	memcpy(newargv + next, *argvp + optind,
					(*argcp+1-optind) * sizeof(char *));
	*argcp += next - optind;
	*argvp = newargv;
	return NULL;
}

/*
 - getanarg - get a malloced argument from the file
 */
static const char *		/* NULL for success, else string literal */
getanarg(f, w, linep)
FILE *f;
struct work *w;
char **linep;			/* where to store pointer if successful */
{
	size_t len;
	char *p;
	char *endp;

	while (w->pending == NULL) {	/* no pending line */
		if ((w->line = of_getline(f, w->buf, sizeof(w->buf))) == NULL)
			return "error in line read";	/* caller checks EOF */
		if (w->line[0] != '#' &&
				*(w->line + strspn(w->line, " \t")) != '\0')
			w->pending = w->line;
	}

	if (w->pending == w->line && w->line[0] != '-') {
		/* fresh plain line */
		w->pending = NULL;
		p = w->line;
		endp = p + strlen(p);
		if (*p == '"' && endp > p+1 && *(endp-1) == '"') {
			p++;
			endp--;
			*endp = '\0';
		}
		if (w->line == w->buf) {
			*linep = malloc(endp - p + 1);
			if (*linep == NULL)
				return "out of memory for new line";
			strcpy(*linep, p);
		} else			/* of_getline already malloced it */
			*linep = p;
		return NULL;
	}

	/* chip off a piece of a pending line */
	p = w->pending;
	p += strspn(p, " \t");
	endp = p + strcspn(p, " \t");
	len = endp - p;
	if (*endp != '\0') {
		*endp++ = '\0';
		endp += strspn(endp, " \t");
	}
	/* endp now points to next real character, or to line-end NUL */
	*linep = malloc(len + 1);
	if (*linep == NULL) {
		if (w->line != w->buf)
			free(w->line);
		return "out of memory for new argument";
	}
	strcpy(*linep, p);
	if (*endp == '\0') {
		w->pending = NULL;
		if (w->line != w->buf)
			free(w->line);
	} else
		w->pending = endp;
	return NULL;
}

/*
 - of_getline - read a line from the file, trim newline off
 */
static char *			/* pointer to line, NULL for eof/error */
of_getline(f, buf, bufsize)
FILE *f;
char *buf;			/* buffer to use, if convenient */
size_t bufsize;			/* size of buf */
{
	size_t len;

	if (fgets(buf, bufsize, f) == NULL)
		return NULL;
	len = strlen(buf);

	if (len < bufsize-1 || buf[bufsize-1] == '\n') {
		/* it fit */
		buf[len-1] = '\0';
		return buf;
	}

	/* oh crud, buffer overflow */
	/* for now, to hell with it */
	return NULL;
}



#ifdef TEST

#include <getopt.h>

char usage[] = "Usage: tester [--foo] [--bar] [--optionsfrom file] arg ...";
struct option opts[] = {
	"foo",		0,	NULL,	'f',
	"bar",		0,	NULL,	'b',
	"builtin",	0,	NULL,	'B',
	"optionsfrom",	1,	NULL,	'+',
	"help",		0,	NULL,	'h',
	"version",	0,	NULL,	'v',
	0,		0,	NULL,	0,
};

int
main(argc, argv)
int argc;
char *argv[];
{
	int opt;
	extern char *optarg;
	extern int optind;
	int errflg = 0;
	const char *p;
	int i;
	FILE *errs = NULL;

	while ((opt = getopt_long(argc, argv, "", opts, NULL)) != EOF)
		switch (opt) {
		case 'f':
		case 'b':
			break;
		case 'B':
			errs = stderr;
			break;
		case '+':	/* optionsfrom */
			p = optionsfrom(optarg, &argc, &argv, optind, errs);
			if (p != NULL) {
				fprintf(stderr, "%s: optionsfrom error: %s\n",
								argv[0], p);
				exit(1);
			}
			break;
		case 'h':	/* help */
			printf("%s\n", usage);
			exit(0);
			break;
		case 'v':	/* version */
			printf("1\n");
			exit(0);
			break;
		case '?':
		default:
			errflg = 1;
			break;
		}
	if (errflg) {
		fprintf(stderr, "%s\n", usage);
		exit(2);
	}

	for (i = 1; i < argc; i++)
		printf("%d: `%s'\n", i, argv[i]);
	exit(0);
}


#endif /* TEST */
